### ISSUE TITLE
Two or more LORAs dont work, error in ComfyUI.

### ISSUE THREAD

**Issue by 3DeeQu** [city96/ComfyUI-GGUF#36]
![Screenshot 2024-08-17 154709](https://github.com/user-attachments/assets/11303670-9c01-431e-b17f-3dae3e4104d2)

**Comment by 3DeeQu** [comment]
Thanks to city96 for active development of the node. The problem was solved after the last update, at least on Q8. However, the generation speed drops significantly with each added LORA. With four LORA, the speed drops x3.

**Comment by city96** [comment]
The speed drop is due to the way LoRA weights are applied. Normally they're pre-applied when you're loading them to the GPU but with a quantized model like this we can only apply them when the weights are dequantized - which only happens when they're being used. This means that for each weight you have to do more and more computations on the fly the more LoRAs you have.

I have some ideas how to speed this up though I suspect it'll always be a bit slower. I'll change the issue title to track the speed issue if that's alright with you.

**Comment by 3DeeQu** [comment]
Yes, of course, and thx for the explanation. If your ideas improve performance, that would be great, since the drop in generation speed by x2-x3 when using 2+ LORA is very depressing... Its much more practical and logical to use FP8 in such a scenario, but if, say, when using 2-4 LORA, the speed drop occurs by... 10-50%, then I would prefer to use Q8 rather than FP8.

**Comment by city96** [comment]
On latest, it's possible to mitigate the slowdown partially by using the advanced loader. The top two options can cause slight changes in the output image, and the bottom option will use slightly more VRAM, so with large/many LoRAs it'll most likely cause OOM issues or overflow into system RAM (slow).

![image](https://github.com/user-attachments/assets/c7651e4b-c2de-4468-9b0d-d4ca5e84ca40)

**Comment by green-anger** [comment]
@city96 Thanks for the update! Could you give a brief summary on how to use it?

1. What are the options for dequant and patch and what are the difference in addition to the mentioned change in the output?
2. If patch_on_device is false will it work like the previous version ignoring the two options above it?

You provided a very good explanation above about why LoRAs slow down the flow when gguf models are used. There're many fp8 models already. I'm trying to better understand all available formats, fp8 are not quantized models and not suffer from the the LoRA issue, is that correct? Compared to gguf models on what level of output quality would you put it, roughly, around Q6_K?

**Comment by city96** [comment]
>Could you give a brief summary on how to use it?

`dequant_dtype` is what the dequantization uses internally. By default it's fp16, with it getting converted to bf16 after the conversion. With `target`, it gets converted to the compute dtype (bf16/fp16) and the dequantization is done in that. This changes the speed the least, and the only real visible difference is with fp32 (though that one will slow it down slightly).

`patch_dtype` changes what precision the LoRA is applied to the weights to. By default comfy calculates the LoRA patches in fp32 since they're only applied once when the model loads to GPU (and with gguf, we calculate them on the GPU when the model runs, so setting them to target (the compute dtype, bf16 or fp16) will cause it to be faster, although the lower precision will cause the image to look different.

`patch_on_device` simply moves the LoRA weights onto the GPU so they don't have to be loaded in one by one from system RAM. It'll increase VRAM usage by whatever the size of the LoRA is.

All 3 options are independent of eachother, so you can mix and match them. The defaults used by the simple/default gguf loader are the defaults in the advanced node.

>Compared to gguf models on what level of output quality would you put it, roughly, around Q6_K?

I have no clue sadly, I think there were changes in comfy that made fp8 better with LoRAs + the inference speed is faster on cards with native FP8 (so RTX 40XX cards and up + quadro/tesla variants). I think the old one I tested was around Q5_K_S at best but at those levels of quantization there's a lot of changes and I never did any proper, large scale A/B testing in this regard.

**Comment by frankyifei** [comment]
I found of way to avoid loading loras at runtime but merge them before start inference. So currently I have to merge loras back to fp32 weight and then convert it back to Q8_0 which I am using now. it needs lot of disk space as well. But does save me lots of time. For example, my currently workflow using ultimate upscaler and HyperSD lora, if loading dynamically, it adds 1.7seconds to each iteration on my old 3060. I am doing 2x upscale, this saves me 55 seconds by using a merged model. 
I am using sd.cpp to convert, it is quite good at this task.

**Comment by ZESTech-hub** [comment]
> I found a way to avoid loading loras at runtime but merge them before start inference. So currently I have to merge loras back to fp32 weight and then convert it back to Q8_0 which I am using now. it needs lot of disk space as well. But does save me lots of time. For example, my currently workflow using ultimate upscaler and HyperSD lora, if loading dynamically, it adds 1.7seconds to each iteration on my old 3060. I am doing 2x upscale, this saves me 55 seconds by using a merged model. I am using sd.cpp to convert, it is quite good at this task.

Could you share the code that you used to transform the model and loras merging?

**Comment by rainlizard** [comment]
I believe the patch_on_device (UNET loader advanced) method city96 mentioned reduces quality slightly.

Someone correct me if I'm wrong, but I think the quality order list for using loras is as follows:

1. 24GB Flux + load lora normally
2. 24GB Flux merged with lora then quantized to 12GB GGUF
3. Quantized 12GB GGUF + load lora normally
4. Quantized 12GB GGUF + load lora normally, using patch_on_device (UNET loader advanced)

The 2nd method is really good, because it's both fast and good quality. frankyifei is likely using it.

Here's the steps to do so:

1. Merge Lora into 24GB safetensors file
2. Convert 24GB safetensors file to 24GB .gguf file
3. Quantize 24GB .gguf file to 12GB .gguf file

It would be cool if there was a custom node which automatically creates a temporary 12GB model file on your hard drive, whenever you change the loras! That would be some crazy automation!

**Comment by frankyifei** [comment]
> I think now that the pip package for gguf has the quantization related code it might be worth adding a node for on-the-fly quantization for the legacy ones (so Q8_0/Q5_1/Q5_0/Q4_1/Q4_0) + a simple save node. That'd make merging in LoRAs or even just quantizing models super easy as long as you don't mind missing out on the K quants.

that would be very useful.

**Comment by KLL535** [comment]
> Падение скорости связано со способом применения весов LoRA. Обычно они применяются заранее, когда вы загружаете их в GPU, но с такой квантованной моделью мы можем применять их только тогда, когда веса деквантуются — что происходит только тогда, когда они используются. Это означает, что для каждого веса вам придется выполнять все больше вычислений на лету, чем больше у вас LoRA.
> 
> У меня есть несколько идей, как ускорить это, хотя я подозреваю, что это всегда будет немного медленнее. Я изменю название проблемы, чтобы отслеживать проблему скорости, если вы не против.

Why then in FORGE with lora everything works quickly regardless of the number of lora? (on NF4). But in comfy-ui NF4 it doesn't work! And why then do we need a gguf if it is slower?

**Comment by LumiOut** [comment]
@frankyifei 

I'm trying to do something similar with Qwen Image Edit 2059 and a custom lora.

I have a 40GB fused Qwen-Image-Edit model (base + Qwen-Image-Lightning + Lora.safetensors) and want to pre-quantize it to W4A4 using DeepCompressor SVDQuant so the LoRAs are permanently baked in — eliminating the 20–25s on-the-fly LoRA conversion delay when loading in ComfyUI.

Done so far:

Fused model into a single .safetensors + config folder
Created a qwen-image-edit-2509.yaml for SVDQuant (W4A4, group_size=64, low_rank_ratio=0.05, mjhq calib dataset)
Output format set to nunchaku for ComfyUI compatibility

I'm not sure that my .yaml is even correct, it's based off the example Flux yaml.

Would you be able to helping me with this?
